revert: rbd: consider remote image health for primary

Madhu-1 opened the pull request:
When the image is force promoted to primary on the cluster the remote image might not be in a replaying
state because due to the split-brain state. This PR reverts back the commit c3c87f2ef33e8d8ad08d7d9f28b59d1aedc4ef31. Which we added to check the remote image status.

closes: #3215 

Signed-off-by: Madhu Rajanna <madhupr007@gmail.com>

Madhu-1 committed:
revert: rbd: consider remote image health for primary

When the image is force promoted to primary on the
cluster the remote image might not be in replaying
state because due to the split brain state. This
PR reverts back the commit
c3c87f2ef33e8d8ad08d7d9f28b59d1aedc4ef31. Which we added
to check the remote image status.

Signed-off-by: Madhu Rajanna <madhupr007@gmail.com>

Madhu-1 commented:
skipping CI as we dont have E2E testing for Mirroring

ShyamsundarR reviewed (dismissed):
LGTM, tested with similar changes as well...

nixpanic reviewed (changes requested):
By reverting this, won't you re-introduce the problem that the change addressed? The original commit points to https://bugzilla.redhat.com/show_bug.cgi?id=2014495.

Can you explain why this is not an issue?

I also expect to see some unit testing for different image/mirroring statuses. This code seems fragile enough that breakage is not easily spotted or detected. The change that introduced the regression was only detected months after is has been merged. Preventing breaking this code in the future should be of the highest priority.

Madhu-1 commented:
> By reverting this, won't you re-introduce the problem that the change addressed? The original commit points to https://bugzilla.redhat.com/show_bug.cgi?id=2014495.
> 

It was not a problem, it's an enhancement we did, as this is getting reverted back I will move the BZ to the next release (downstream )

> Can you explain why this is not an issue?
>

As I mentioned above its an enhancement we tried to due but introduced a regression in a corner case. AFAIK, so revert back is fine 
 
> I also expect to see some unit testing for different image/mirroring statuses. This code seems fragile enough that breakage is not easily spotted or detected. 

Unit testing is not possible as `checkHealthyPrimary` makes a ceph call to get the mirroring image details.

> The change that introduced the regression was only detected months after is has been merged. Preventing breaking this code in the future should be of the highest priority.

not all the cases can be covered in the happy path testing, as you know testing DR is always hard we get new bugs in different areas when we started exploring different areas.

nixpanic reviewed (commented):
> > By reverting this, won't you re-introduce the problem that the change addressed? The original commit points to https://bugzilla.redhat.com/show_bug.cgi?id=2014495.
> 
> It was not a problem, it's an enhancement we did, as this is getting reverted back I will move the BZ to the next release (downstream )

Ok! It would be good to have the enhancement documented in this project too. Please create an issue for that.

> > Can you explain why this is not an issue?
> 
> As I mentioned above its an enhancement we tried to due but introduced a regression in a corner case. AFAIK, so revert back is fine

Thanks for considering, that was not clear to me by reading the PR description or commit message.

> > I also expect to see some unit testing for different image/mirroring statuses. This code seems fragile enough that breakage is not easily spotted or detected.
> 
> Unit testing is not possible as `checkHealthyPrimary` makes a ceph call to get the mirroring image details.

It is important to find a way for testing this, at least minimally. A regression was introduced because of missing tests, and there is no guarantee that a future _enhancement_ like this will not break things again.

> > The change that introduced the regression was only detected months after is has been merged. Preventing breaking this code in the future should be of the highest priority.
> 
> not all the cases can be covered in the happy path testing, as you know testing DR is always hard we get new bugs in different areas when we started exploring different areas.

Unit testing is not only about happy-path testing. Specially for features that can not be tested in e2e (yet), unit-tests are important. Can you consider re-structuring the code a bit, so that the mirror status and planned actions can be validated?

Madhu-1 committed:
rbd: add unit test for checkHealthyPrimary

Removed the code in checkHealthyPrimary which
makes the ceph call, passing it as input now.
Added unit test for checkHealthyPrimary function

Signed-off-by: Madhu Rajanna <madhupr007@gmail.com>

nixpanic reviewed (dismissed):
much appreciated, thanks!

Madhu-1 commented:
> much appreciated, thanks!

Fixed CI test failure, PTAL

Madhu-1 committed:
rbd: fix checkHealthyPrimary to consider up+stopped state

we need to check for image should be in up+stopped state
not anyone of the state for that the we need to use
OR check not the AND check.

Signed-off-by: Madhu Rajanna <madhupr007@gmail.com>

nixpanic commented:
@mergifyio requeue

Madhu-1 commented:
@mergifyio refresh
